File: typescript/src/storage/summarizingChatStorage.ts
import { ChatStorage } from "./chatStorage";
import { ConversationMessage } from "../types";

/**
 * A callable that compresses a conversation history.
 *
 * Receives the full history and the number of recent pairs to keep verbatim.
 * Must return the compressed history (summary message(s) + recent pairs).
 */
export type ChatSummarizer = (
  history: ConversationMessage[],
  keepLast: number
) => Promise<ConversationMessage[]>;

/**
 * A `ChatStorage` wrapper that automatically compresses long conversation histories.
 *
 * Wraps any `ChatStorage` implementation. On every `fetchChat` call, if the
 * returned history exceeds `triggerAt` message pairs the user-supplied
 * `summarizer` callable is invoked. The compressed result is cached internally
 * so subsequent fetches are fast without re-calling the summarizer.
 *
 * `fetchAllChats` is never intercepted — the classifier always sees the full
 * cross-agent history unmodified.
 *
 * @example
 * ```typescript
 * import { SummarizingChatStorage, InMemoryChatStorage } from 'agent-squad';
 *
 * async function mySummarizer(history, keepLast) {
 *   const old = history.slice(0, -keepLast * 2);
 *   const recent = history.slice(-keepLast * 2);
 *   const summary = await callLlmToSummarize(old);
 *   return [
 *     { role: 'user', content: [{ text: `[Summary]: ${summary}` }] },
 *     ...recent,
 *   ];
 * }
 *
 * const storage = new SummarizingChatStorage(
 *   new InMemoryChatStorage(),
 *   mySummarizer,
 *   20,  // triggerAt: summarize when history exceeds 20 pairs
 *   2,   // keepLast: keep the 2 most recent pairs verbatim
 * );
 * ```
 */
export class SummarizingChatStorage extends ChatStorage {
  /**
   * Internal cache: key → compressed history.
   * After summarization the compressed result is stored here so subsequent
   * `fetchChat` calls return immediately without hitting the inner storage or
   * re-invoking the summarizer. A `saveChatMessage` call invalidates the entry
   * so the next fetch re-evaluates from the inner store.
   */
  private readonly cache: Map<string, ConversationMessage[]> = new Map();

  /**
   * @param storage  The inner `ChatStorage` to wrap.
   * @param summarizer  Async callable that compresses history.
   * @param triggerAt  Number of message **pairs** above which summarization
   *   is triggered. Defaults to 20.
   * @param keepLast  Number of most-recent message pairs to keep verbatim.
   *   Passed to the summarizer as the second argument. Defaults to 2.
   */
  constructor(
    private readonly storage: ChatStorage,
    private readonly summarizer: ChatSummarizer,
    private readonly triggerAt: number = 20,
    private readonly keepLast: number = 2
  ) {
    super();
  }

  async saveChatMessage(
    userId: string,
    sessionId: string,
    agentId: string,
    newMessage: ConversationMessage,
    maxHistorySize?: number
  ): Promise<ConversationMessage[]> {
    // Invalidate cache so the next fetchChat re-evaluates from the inner store.
    this.cache.delete(this.cacheKey(userId, sessionId, agentId, maxHistorySize));
    return this.storage.saveChatMessage(userId, sessionId, agentId, newMessage, maxHistorySize);
  }

  async fetchChat(
    userId: string,
    sessionId: string,
    agentId: string,
    maxHistorySize?: number
  ): Promise<ConversationMessage[]> {
    const key = this.cacheKey(userId, sessionId, agentId, maxHistorySize);

    // Return cached compressed history if available.
    const cached = this.cache.get(key);
    if (cached !== undefined) {
      return cached;
    }

    const history = await this.storage.fetchChat(userId, sessionId, agentId, maxHistorySize);

    if (history.length > this.triggerAt * 2) {
      const compressed = await this.summarizer(history, this.keepLast);
      this.cache.set(key, compressed);
      return compressed;
    }

    return history;
  }

  async fetchAllChats(
    userId: string,
    sessionId: string
  ): Promise<ConversationMessage[]> {
    // Never intercepted — classifier must see the full cross-agent history.
    return this.storage.fetchAllChats(userId, sessionId);
  }

  private cacheKey(userId: string, sessionId: string, agentId: string, maxHistorySize?: number): string {
    return `${userId}#${sessionId}#${agentId}#${maxHistorySize ?? 'nil'}`;
  }
}

File: python/src/agent_squad/storage/summarizing_chat_storage.py
"""
SummarizingChatStorage — a ChatStorage wrapper that compresses long conversation
histories using a user-supplied async callable.

When the history for a given agent exceeds ``trigger_at`` message pairs, the
wrapper calls the summarizer, writes the compressed result back to the inner
store (so subsequent fetches are fast), and returns the compressed history.

``fetch_all_chats`` is never intercepted — the classifier always sees the full
cross-agent picture unmodified.

Usage::

    from agent_squad.storage import SummarizingChatStorage, InMemoryChatStorage
    from agent_squad.types import ConversationMessage, ParticipantRole

    async def my_summarizer(
        history: list[ConversationMessage],
        keep_last: int,
    ) -> list[ConversationMessage]:
        old = history[:-keep_last * 2]
        recent = history[-keep_last * 2:]
        summary_text = await call_llm_to_summarize(old)
        return [
            ConversationMessage(
                role=ParticipantRole.USER.value,
                content=[{"text": f"[Conversation summary]: {summary_text}"}],
            )
        ] + recent

    storage = SummarizingChatStorage(
        storage=InMemoryChatStorage(),
        summarizer=my_summarizer,
        trigger_at=20,   # summarize when history exceeds 20 pairs
        keep_last=2,     # keep the 2 most recent pairs verbatim
    )
"""

from typing import Callable, Awaitable, Optional, Union

from agent_squad.storage.chat_storage import ChatStorage
from agent_squad.types import ConversationMessage, TimestampedMessage


class SummarizingChatStorage(ChatStorage):
    """A ``ChatStorage`` wrapper that automatically compresses long histories.

    Wraps any ``ChatStorage`` implementation. On every ``fetch_chat`` call,
    if the returned history exceeds ``trigger_at`` message pairs the summarizer
    callable is invoked, the compressed result is written back to the inner
    store, and the compressed history is returned to the caller.

    All other methods (``save_chat_message``, ``save_chat_messages``,
    ``fetch_all_chats``) are pure delegations to the inner storage.

    Args:
        storage: The inner ``ChatStorage`` to wrap.
        summarizer: An async callable with the signature
            ``async (history: list[ConversationMessage], keep_last: int)
            -> list[ConversationMessage]``.
            Receives the full history and the number of recent pairs to
            preserve verbatim. Must return the compressed history.
        trigger_at: Number of message **pairs** (user + assistant = 1 pair)
            above which summarization is triggered. Default: 20.
        keep_last: Number of most-recent message pairs to keep verbatim and
            pass to the summarizer as context. Default: 2.
    """

    def __init__(
        self,
        storage: ChatStorage,
        summarizer: Callable[[list[ConversationMessage], int], Awaitable[list[ConversationMessage]]],
        trigger_at: int = 20,
        keep_last: int = 2,
    ) -> None:
        super().__init__()
        self._storage = storage
        self._summarizer = summarizer
        self._trigger_at = trigger_at
        self._keep_last = keep_last
        # Internal cache: key → compressed history.
        # After summarization the compressed result is stored here so subsequent
        # fetch_chat calls return immediately without hitting the inner storage or
        # re-invoking the summarizer. A save_chat_message call invalidates the entry
        # so the next fetch re-evaluates from the inner store.
        self._cache: dict[str, list[ConversationMessage]] = {}

    @staticmethod
    def _cache_key(user_id: str, session_id: str, agent_id: str) -> str:
        return f"{user_id}#{session_id}#{agent_id}"

    async def save_chat_message(
        self,
        user_id: str,
        session_id: str,
        agent_id: str,
        new_message: Union[ConversationMessage, TimestampedMessage],
        max_history_size: Optional[int] = None,
    ) -> bool:
        # Invalidate cache so the next fetch_chat re-evaluates from the inner store.
        self._cache.pop(self._cache_key(user_id, session_id, agent_id), None)
        return await self._storage.save_chat_message(
            user_id, session_id, agent_id, new_message, max_history_size
        )

    async def save_chat_messages(
        self,
        user_id: str,
        session_id: str,
        agent_id: str,
        new_messages: Union[list[ConversationMessage], list[TimestampedMessage]],
        max_history_size: Optional[int] = None,
    ) -> bool:
        # Invalidate cache so the next fetch_chat re-evaluates from the inner store.
        self._cache.pop(self._cache_key(user_id, session_id, agent_id), None)
        return await self._storage.save_chat_messages(
            user_id, session_id, agent_id, new_messages, max_history_size
        )

    async def fetch_chat(
        self,
        user_id: str,
        session_id: str,
        agent_id: str,
        max_history_size: Optional[int] = None,
    ) -> list[ConversationMessage]:
        key = self._cache_key(user_id, session_id, agent_id)

        # Return cached compressed history if available.
        if key in self._cache:
            return self._cache[key]

        # Fetch the full history — we need it untruncated to evaluate the
        # trigger threshold. max_history_size is accepted for interface
        # compatibility but size management is delegated to trigger_at/keep_last.
        history = await self._storage.fetch_chat(user_id, session_id, agent_id)

        if len(history) > self._trigger_at * 2:
            compressed = await self._summarizer(history, self._keep_last)
            # Cache the compressed result — subsequent fetches return this directly.
            self._cache[key] = compressed
            return compressed

        return history

    async def fetch_all_chats(
        self,
        user_id: str,
        session_id: str,
    ) -> list[ConversationMessage]:
        # Never intercepted — the classifier must see the full cross-agent history.
        return await self._storage.fetch_all_chats(user_id, session_id)
